File: euler23.py
import numpy as np

def d(x):
  r = 1
  if x <= 1:
    return 0
  else:
    sq = x**(.5)
    if int(sq) == 1:
      return 1;
    for i in xrange(2,int(sq)):
      if x % i == 0:
        r += i + x / i
    if sq == int(sq):
      r += sq
    elif x % int(sq) == 0:
      r += int(sq) + x / int(sq)
    return r


def find_nonabundant_sums(cap,num_abundant):
  ab_nums = np.zeros(num_abundant)
  curr_ind = 0
  mask = np.arange(cap)
#  print mask[345];
  s = 0
#  print d(12)
  for x in xrange(cap):
    divs = d(x)
    if x < divs:
      ab_nums[curr_ind] = x
      curr_ind += 1
      for j in xrange(curr_ind):
        s = x + ab_nums[j]
        if s < cap:
 #         print ab_nums
          mask[s] = 0
  print np.sum(mask)
  print curr_ind
#  print mask[(mask > 0)]
  
      
#foo = np.arange(25)
#print np.sum(foo) - 24
print d(900)    
find_nonabundant_sums(2000,10000)

File: euler23.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
//#define CAP 28132
#define CAP 20162
#define AB_CAP 10000

long int d(int x);

int main()
{
 int mask[CAP];
 int ab_nums[AB_CAP];
 int curr_ind = 0;
 long int i;
 unsigned long long int sum; 
 memset(mask,0,sizeof(mask));
 
 for (i = 0; i < CAP; i++)
 {
   if (d(i) > i)
   {
     ab_nums[curr_ind++] = i;
     int j;
     for (j = 0; j < curr_ind; j++)
     {
       long int s = i + ab_nums[j];
       if (s < CAP) mask[s] = 1;
     }
   }
   
   if (!mask[i]) sum += i;
 }
 
 printf("Sum: %llu\n",sum);
// printf("d(900) %ld\n",d(12));
}

long int d(int x)
{
  if (x < 2) return 0; 
  long int sum = 1;
  int sq = (int)sqrt(x);
  if (sq == 1) return 1;
  int i;
  for (i = 2; i < sq; i++)
  {
    int mod = x % i;
    int div = x / i;
    if (mod == 0)
    {
      sum += i + div;
    }    
  }
  
  if (sq*sq == x) sum += sq;
  else if (x % sq == 0) sum += sq + x / sq;
  return sum;
}